### ISSUE TITLE
com.hazelcast.map.impl.query.QueryIndexingTest [HZ-2491]

### ISSUE THREAD

**Issue by Patras3** [hazelcast/hazelcast#23645]
_master_ (commit cf062b9b7d44fae04ceb31b2b94e52a3d540219f)

Failed on oracle-11: https://jenkins.hazelcast.com/job/Hazelcast-master-sonar-new/44/testReport/junit/com.hazelcast.map.impl.query/QueryIndexingTest/
for two tests:
 `testResultsHaveNullFields_whenPredicateTestsForNull`
`testResultsHaveNullFields_whenUsingIndexes` with the same NPE.
<details><summary>Stacktrace:</summary>

```
java.lang.NullPointerException
	at com.hazelcast.map.impl.query.QueryIndexingTest.tearDown(QueryIndexingTest.java:81)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at com.hazelcast.test.AbstractHazelcastClassRunner$ThreadDumpAwareRunAfters.evaluate(AbstractHazelcastClassRunner.java:381)
	at com.hazelcast.test.DumpBuildInfoOnFailureRule$1.evaluate(DumpBuildInfoOnFailureRule.java:37)
	at com.hazelcast.test.jitter.JitterRule$1.evaluate(JitterRule.java:104)
	at com.hazelcast.test.metrics.MetricsRule$1.evaluate(MetricsRule.java:63)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at com.hazelcast.test.HazelcastParallelClassRunner.access$301(HazelcastParallelClassRunner.java:49)
	at com.hazelcast.test.HazelcastParallelClassRunner$TestRunner.run(HazelcastParallelClassRunner.java:162)
	at java.base/java.lang.Thread.run(Thread.java:834)

```
</details>

Standard output can be found here - https://s3.console.aws.amazon.com/s3/buckets/j-artifacts/Hazelcast-master-sonar-new/44/

**Comment by olukas** [comment]
Failed again on _master_ with oracle-11: https://jenkins.hazelcast.com/view/Official%20Builds/job/Hazelcast-master-sonar-new/47/testReport/com.hazelcast.map.impl.query/QueryIndexingTest/

**Comment by sumnerib** [comment]
Failed again on _master_ with oracle-11: https://jenkins.hazelcast.com/job/Hazelcast-master-sonar-new/55/testReport/junit/com.hazelcast.map.impl.query/QueryIndexingTest/

**Comment by olukas** [comment]
Failed again on _master_ with oracle-11: https://jenkins.hazelcast.com/view/Official%20Builds/job/Hazelcast-master-sonar-new/61/testReport/com.hazelcast.map.impl.query/QueryIndexingTest/

**Comment by olukas** [comment]
Fails constantly on Sonar build.

**Comment by JamesHazelcast** [comment]
This issue is no longer a `constant-test-failure`, but the most recent build of the `Hazelcast master sonar` (oracle-11) job failed with the same failures, after passing for a few in a row: https://jenkins.hazelcast.com/job/Hazelcast-master-sonar/140/testReport/junit/com.hazelcast.map.impl.query/QueryIndexingTest/

Looking at the output of this failure, it appears to be another Mockito related failure:
```
[ERROR] com.hazelcast.map.impl.query.QueryIndexingTest.testResultsHaveNullFields_whenPredicateTestsForNull  Time elapsed: 7.092 s  <<< ERROR!
java.lang.NoClassDefFoundError: Could not initialize class com.hazelcast.query.SampleTestObjects$Employee$MockitoMock$p9OzJdhv
	at jdk.internal.reflect.GeneratedSerializationConstructorAccessor9466.newInstance(Unknown Source)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
	at org.objenesis.instantiator.sun.SunReflectionFactoryInstantiator.newInstance(SunReflectionFactoryInstantiator.java:48)
	at org.objenesis.ObjenesisBase.newInstance(ObjenesisBase.java:73)
	at org.mockito.internal.creation.instance.ObjenesisInstantiator.newInstance(ObjenesisInstantiator.java:22)
	at org.mockito.internal.creation.bytebuddy.InlineDelegateByteBuddyMockMaker.doCreateMock(InlineDelegateByteBuddyMockMaker.java:376)
	at org.mockito.internal.creation.bytebuddy.InlineDelegateByteBuddyMockMaker.createMock(InlineDelegateByteBuddyMockMaker.java:334)
	at org.mockito.internal.creation.bytebuddy.InlineByteBuddyMockMaker.createMock(InlineByteBuddyMockMaker.java:56)
	at org.mockito.internal.util.MockUtil.lambda$createMock$1(MockUtil.java:94)
	at java.base/java.util.Optional.orElseGet(Optional.java:369)
	at org.mockito.internal.util.MockUtil.createMock(MockUtil.java:92)
	at org.mockito.internal.MockitoCore.mock(MockitoCore.java:88)
	at org.mockito.Mockito.mock(Mockito.java:2101)
	at com.hazelcast.mock.MockUtil.serializableSpy(MockUtil.java:43)
	at com.hazelcast.map.impl.query.QueryIndexingTest.newEmployees(QueryIndexingTest.java:123)
	at com.hazelcast.map.impl.query.QueryIndexingTest.setUp(QueryIndexingTest.java:66)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.RunBefores.invokeMethod(RunBefores.java:33)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:24)
	at com.hazelcast.test.AbstractHazelcastClassRunner$ThreadDumpAwareRunAfters.evaluate(AbstractHazelcastClassRunner.java:365)
	at com.hazelcast.test.DumpBuildInfoOnFailureRule$1.evaluate(DumpBuildInfoOnFailureRule.java:37)
	at com.hazelcast.test.jitter.JitterRule$1.evaluate(JitterRule.java:104)
	at com.hazelcast.test.metrics.MetricsRule$1.evaluate(MetricsRule.java:63)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at com.hazelcast.test.HazelcastParallelClassRunner.access$101(HazelcastParallelClassRunner.java:49)
	at com.hazelcast.test.HazelcastParallelClassRunner$TestRunner.run(HazelcastParallelClassRunner.java:162)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

I will follow up with @arodionov to see if he has any insights on this before investigating further.

**Comment by arodionov** [comment]
The current issue https://github.com/hazelcast/hazelcast/issues/23645#issuecomment-1564243903 has a different stack trace and another failure type.

**Comment by JamesHazelcast** [comment]
> The current issue [#23645 (comment)](https://github.com/hazelcast/hazelcast/issues/23645#issuecomment-1564243903) has a different stack trace and another failure type.

The NPE is caused on `tearDown()`, which (at least in the case of the most recent failure) is caused by the `setUp()` failing to complete (due to ClassNotFound when trying to instantiate the mocked `Employee` object), which means the `nodeFactory` object is not created and therefore null when `nodeFactory.shutdownAll()` is called in `tearDown()`. The NPEs are present in both failures, and if I were a betting man I'd put my money on it being the same underlying cause (although not provable for the first failure due to the loss of console output and lack of detail in the surefire reports)

**Comment by AleksPeychev** [comment]
Failed again on _5.3.8_ with oracle-17: https://jenkins.hazelcast.com/view/Security%20and%20Compliance/job/Hazelcast-sonar-custom-branch/38/testReport/junit/com.hazelcast.map.impl.query/QueryIndexingTest/testResultsHaveNullFields_whenPredicateTestsForNull_2/

**Comment by olukas** [comment]
Closing, it seems to be fixed on stream 5.4.z and newer. We should not actively reopen TF for older versions.
